File: Java/chemistrytool/concentration/Concentration.java
package chemistrytool.concentration;

import chemistrytool.util.Exceptions.ElementNotFoundInCompoundException;
import chemistrytool.util.Exceptions.MissingInputException;
import chemistrytool.util.JSONTool;
import chemistrytool.util.Parser;

public class Concentration {

    /**
     * Concentration class contains functions to calculate unknown concentration of substance.
     *  Example:  Concentration concentration = new Concentration("10kg","12millilitre","CH4");
     *            concentration.CalculateMolarity();   . . . .  Molarity = given mass / (Molar mass * volume of solution)
     */

    Parser p = new Parser();
    private String given1;
    private String given2;
    private String given3 = "";

    public Concentration(String given1) {
        this.given1 = given1;
    }

    public Concentration(String given1, String given2) {
        this.given1 = given1;
        this.given2 = given2;
    }

    public Concentration(String given1, String given2, String given3) {
        this.given1 = given1;
        this.given2 = given2;
        this.given3 = given3;
    }

    /**
     * Returns value that express unknown concentration of a substance in terms of molarity.
     * @return
     * @throws MissingInputException
     */

    public double CalculateMolarity()throws MissingInputException {
        double molarity = 0, mass = 0,mole = 0,volume = 0,molar_mass = 0;
        String[] givens = {given1,given2,given3};
        for(String given : givens) {
            if (p.isMass(given)) {
                mass = p.parse(given);
                if (given.contains("kg")){
                    mass = mass * 1000;
                }
            }else if(p.isMole(given)){
                mole = p.parse(given);
            }else if(p.isVolume(given)){
                volume = p.parse(given);
                if (given.contains("millilitre")){
                    volume = volume / 1000;
                }
            }else{
                molar_mass = p.getMolarMass(p.getElements(given),given);
            }
        }
        if(mass == 0 && molar_mass == 0){
            molarity = mole/volume;
        }else if(mole == 0){
            molarity = mass / (molar_mass * volume);
        }else{
            throw new MissingInputException("Missing input!");
        }
        return molarity;
    }

    /**
     * Returns value that express unknown concentration of a substance in terms of molality.
     * @return
     * @throws MissingInputException
     */

    public double CalculateMolality()throws MissingInputException {
        double molality = 0, mass_of_solute = 0,mole = 0,mass_of_solvent = 0,molar_mass = 0;
        String[] givens = {given1,given2,given3};
        for(String given : givens) {
            if (p.isMass(given)) {
                if (given.contains("solvent")){
                    mass_of_solvent = p.parse(given);
                    if (given.contains("gram")){
                        mass_of_solvent = mass_of_solvent/1000;
                    }
                }else {
                    mass_of_solute = p.parse(given);
                }
            }else if(p.isMole(given)){
                mole = p.parse(given);
            }else{
                molar_mass = p.getMolarMass(p.getElements(given),given);
            }
        }
        System.out.println(molar_mass);
        if(mass_of_solute == 0 && molar_mass == 0){
            molality = mole/mass_of_solvent;
        }else if(mole == 0){
            molality = mass_of_solute / (molar_mass * mass_of_solvent);
        }else{
            throw new MissingInputException("Missing input!");
        }
        return molality;
    }

    /**
     * Returns value that express unknown concentration of a substance in terms of normality.
     * @return
     * @throws MissingInputException
     */

    public double CalculateNormality()throws MissingInputException {
        double normality = 0, mass = 0,volume = 0,number_of_equivalent_mass = 0,equivalent_mass = 0;
        String[] givens = {given1,given2,given3};
        for(String given : givens) {
            if (p.isMass(given)) {
                mass = p.parse(given);
            }else if(p.isVolume(given)){
                volume = p.parse(given);
                if (given.contains("millilitre")){
                    volume = volume / 1000;
                }
            }else if (p.isNumberOfEquivalent(given)){
                number_of_equivalent_mass = p.parse(given);
            }else{
                equivalent_mass = p.CalculateEquivalentMass(given);
            }
        }
        if(mass == 0 && equivalent_mass == 0){
            normality = number_of_equivalent_mass/volume;
        }else if(number_of_equivalent_mass == 0){
            normality = mass / (equivalent_mass * volume);
        }else{
            throw new MissingInputException("Missing input!");
        }
        return normality;
    }


    /**
     * Calculates amount of element in a compound.
     * @throws ElementNotFoundInCompoundException
     */

    public String CalculatePercentageComposition()throws ElementNotFoundInCompoundException {
        String compound = given2,element = given1;
        if (compound.contains(element)) {
            int coefficient = Character.isDigit(compound.charAt(0)) ? compound.charAt(0) - 48 : 1;
            double percentage_composition = (p.getMolarMass(p.getElements(element), element) / p.getMolarMass(p.getElements(compound), compound)) * 100;
            return "There is " + percentage_composition + "% of " + element + " in " + compound;
        }else{
            throw new ElementNotFoundInCompoundException("Element not found in the given compound!");
        }
    }

    /**
     *  Returns the electronic configuration of an element.
     */

    public String getElectronicConfiguration(){
        JSONTool js = new JSONTool();
        return String.valueOf(js.decode(given1).get("electron_configuration"));
    }




}

File: Python/Concentration.py
from util.Parser import *


class Concentration:

    """Concentration class contains functions to calculate unknown concentration of substance.
       Example:  concentration = Concentration("10kg","12millilitre","CH4")
                 concentration.calculate_molarity()   . . . .  Molarity = given mass / (Molar mass * volume of solution) """

    def __init__(self,given1,given2,given3 = None):
        self.given1 = given1
        self.given2 = given2
        self.given3 = given3
        self.molarity = 0
        self.molality = 0
        self.normality = 0
        self.mass = 0
        self.mole = 0
        self.volume = 0
        self.molar_mass = 0
        self.mass_of_solute = 0
        self.mass_of_solvent = 0
        self.number_of_equivalent = 0
        self.equivalent_mass = 0
        self.givens = [self.given1,self.given2,self.given3]

    def calculate_molarity(self):

        """ Returns value that express unknown concentration of a substance in terms of molarity."""

        for given in self.givens:
            if given is None:
                pass
            else:
                if is_mass(given):
                    self.mass = get_value(given)
                    if given.endswith('kg'):
                        self.mass = self.mass * 1000
                elif is_mole(given):
                    self.mole = get_value(given)
                elif is_volume(given):
                    self.volume = get_value(given)
                    if given.endswith('millilitre'):
                        self.volume = self.volume / 1000
                else:
                    self.molar_mass = calculate_molar_mass(get_elements(given), given)
        if self.mass == 0 and self.molar_mass == 0:
            self.molarity = self.mole/self.volume
        elif self.mole == 0:
            self.molarity = self.mass/self.volume * self.molar_mass
        else:
            print('Missing Input')
        return self.molarity

    def calculate_molality(self):

        """ Returns value that express unknown concentration of a substance in terms of molality."""

        for given in self.givens:
            if given is None:
                pass
            else:
                if is_mass(given):
                    self.mass_of_solute = get_value(given)
                    if given.endswith('g'):
                        self.mass = self.mass / 1000
                elif given.endswith('solvent'):
                    self.mass_of_solvent = get_value(given)
                elif is_mole(given):
                    self.mole = get_value(given)
                else:
                    self.molar_mass = calculate_molar_mass(get_elements(given), given)
        if self.mass_of_solute == 0 and self.molar_mass == 0:
            self.molality = self.mole/self.mass_of_solvent
        elif self.mole == 0:
            self.molality = self.mass_of_solute / (self.molar_mass * self.mass_of_solvent)
        else:
            print('Missing Input')
        return self.molality

    def calculate_normality(self):

        """Returns value that express unknown concentration of a substance in terms of normality."""

        for given in self.givens:
            if given is None:
                pass
            else:
                if is_volume(given):
                    self.volume = get_value(given)
                    if given.endswith('millilitre'):
                        self.volume = self.volume / 1000
                elif is_number_of_equivalent_mass(given):
                    self.number_of_equivalent = get_value(given)
                elif is_mass(given):
                    self.mass = get_value(given)
                else:
                    self.equivalent_mass = calculate_equivalent_mass(given)

        if self.mass == 0 and self.equivalent_mass == 0:
            self.normality = self.number_of_equivalent/self.volume
        elif self.number_of_equivalent == 0:
            self.normality = self.mass / self.equivalent_mass * self.volume
        else:
            print('Missing Input')
        return self.normality

    def calculate_percentage_composition(self):

        """ Calculates amount of element in a compound. """

        molar_mass = calculate_molar_mass(get_elements(self.given1),self.given1)
        molar_mass2 = calculate_molar_mass(get_elements(self.given2),self.given2)
        return "There is " + str((molar_mass/molar_mass2) * 100) + " % " + self.given1 + " in " + self.given2









